lib/helpers/positional_generator.rb

```
@@ -0,0 +1,475 @@
+    #       g_.digit(length: 5)
+    #       g_.lit("/")
+    #     end
+    #   end
+    def oneof(name: nil, &block)
+      @components << Component.new(@components.count, name, [], Oneof.new(self, block))
+    end
+
+    ##
+    # A group of generators. Useful for {#oneof}.
+    #
+    # @param name [Symbol] the name for this node in the group
+    # @param block [Method] a subgenerator block
+    # @return [void]
+    def group(name: nil, &block)
+      @components << Component.new(@components.count, name, [], Group.new(@as_type, block))
+    end
+
+    ##
+    # Generate the value.
+    #
+    # @return [String] if +as_type+ is +:string+
+    def build
+      graph = build_graph
+      stack = build_stack(graph)
+      values = generate_values(stack)
+      convert(values)
+    end
+
+    private
+
+    ##
+    # Turn the components into a graph following dependencies.
+    #
+    # @return [Array<(Integer, Integer)>]
+    #
+    # Components can have dependencies. Here's one where a computation (b)
+    # depends on a value generated after it (c):
+    #
+    #     @components = [
+    #       Int.new(0, :a, 1, nil),
+    #       Computed.new(1, :b, [:c]) { |c| c + 1 },
+    #       Int.new(2, :c, 1, nil),
+    #     ]
+    #
+    # We can think of a graph like so:
+    #
+    #      (a)  (c)
+    #       |    |
+    #       |   (b)
+    #       \   /
+    #        end
+    #
+    # Or in Mermaid:
+    #
+    # ```mermaid
+    # stateDiagram-v2
+    #     a --> [*]
+    #     c --> b
+    #     b --> [*]
+    # ```
+    #
+    # This method builds that graph, using their positional locations as the
+    # ID. The end state is represented as +nil+. So continuing the example
+    # above, it will give this output:
+    #
+    #     [
+    #       [0, nil],
+    #       [2, 1],
+    #       [1, nil],
+    #     ]
+    #
+    # Later we can look up the appropriate component by indexing into the
+    # +@components+ array.
+    def build_graph
```

Interesting approach! 

Is there a need to also detect any cycles in the dependency graph? I guess if you're keeping track of `seen` nodes, you would be able to detect it, and maybe raise an error?

Not sure it would be possible to generate all values for these component if there's a dependency cycle, so maybe raising "cycle detected etc." when the cycle is detected could be useful to the user.

By the way, have you considered generating a topological sort of the graph and calculating/generating values based on that? I know there's a gem called `tsort`, but I've never used it.

---

> Is there a need to also detect any cycles in the dependency graph?

Good question, and I'm not sure!

This is an internal class that ideally the user will not control, so hopefully faker devs and people extending faker will be able to wield it responsibly. Currently there is no concept of generating these graph nodes dynamically; they are defined once, and that's the result every time.

That said, it will be a nice debugging tool to catch those cycles. What do you think, should we build that in?

> generating a topological sort of the graph

Oh hey, those are the words I should have searched for! I guess I did that manually. Would you rather I bring in the tsort gem?

---

I don't think it's necessary to use tsort, your implementation is simple enough to follow. It was just a thought I had when looking at the code. And maybe we don't need to worry about cycle detection right now. 

I'm happy with the current state. Let me know if you want to add anything else, or else I'll just merge it. Thanks!

---

All good from my end. Thank you so much for taking the time to read it and accept it into your repo!